File: backend/Application/Services/AutomationWebhookEngine.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using ITIncidentCopilot.Api.Entities;

namespace ITIncidentCopilot.Api.Application.Services
{
    public class WebhookExecutionLog
    {
        public Guid ExecutionId { get; set; } = Guid.NewGuid();
        public string EventType { get; set; } = string.Empty;
        public string RuleName { get; set; } = string.Empty;
        public string TargetEndpoint { get; set; } = "http://localhost:5678/webhook/incident-events";
        public string PayloadJson { get; set; } = "{}";
        public string Status { get; set; } = "DISPATCHED_200_OK";
        public DateTime Timestamp { get; set; } = DateTime.UtcNow;
    }

    public interface IAutomationWebhookEngine
    {
        Task<WebhookExecutionLog> DispatchWebhookAsync(string eventType, Incident incident, string actionTarget);
    }

    public class AutomationWebhookEngine : IAutomationWebhookEngine
    {
        private const string WebhookSecret = "n8n_sec_copilot_webhook_key_2026";

        public Task<WebhookExecutionLog> DispatchWebhookAsync(string eventType, Incident incident, string actionTarget)
        {
            string payload = $"{{\"event\":\"{eventType}\",\"ticket\":\"{incident.TicketNumber}\",\"severity\":\"{incident.Severity}\",\"action\":\"{actionTarget}\",\"timestamp\":\"{DateTime.UtcNow:o}\"}}";
            string hmacSignature = ComputeHmacSha256(payload, WebhookSecret);

            return Task.FromResult(new WebhookExecutionLog
            {
                EventType = eventType,
                RuleName = GetRuleNameForEvent(eventType),
                TargetEndpoint = "http://localhost:5678/webhook/incident-events",
                PayloadJson = payload,
                Status = $"DISPATCHED_200_OK (HMAC: {hmacSignature.Substring(0, 10)}...)",
                Timestamp = DateTime.UtcNow
            });
        }

        private string GetRuleNameForEvent(string eventType) => eventType switch
        {
            "CRITICAL_P1_ALERT" => "Notify IT Manager (P1 Critical Alert)",
            "SLA_UNACKNOWLEDGED_ESCALATION" => "Escalate Unacknowledged Ticket (>15m)",
            "TICKET_RESOLVED_NOTIFY_USER" => "Send Resolution Email to User",
            _ => "Generic Automation Trigger"
        };

        private string ComputeHmacSha256(string rawData, string secret)
        {
            using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
            byte[] hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(rawData));
            return Convert.ToHexString(hash);
        }
    }
}

File: src/services/automationWebhookEngine.ts
import type { Incident } from '../types';

export interface WebhookLogEntry {
  id: string;
  eventType: 'CRITICAL_P1_ALERT' | 'SLA_UNACKNOWLEDGED_ESCALATION' | 'TICKET_RESOLVED_NOTIFY_USER';
  ruleName: string;
  targetEndpoint: string;
  payloadJson: string;
  hmacSignature: string;
  status: string;
  timestamp: string;
}

export class AutomationWebhookEngine {
  private static liveStreamLogs: WebhookLogEntry[] = [
    {
      id: 'wh-101',
      eventType: 'CRITICAL_P1_ALERT',
      ruleName: 'Notify IT Manager (P1 Critical Alert)',
      targetEndpoint: 'http://localhost:5678/webhook/p1-critical-alert',
      payloadJson: '{"event":"CRITICAL_P1_ALERT","ticket":"INC-2026-8812","severity":"CRITICAL","impact":"Executive Print Queue"}',
      hmacSignature: 'sha256=8f9a2e1d7c3b4a5f6e8d0c2b4a6f8e0d',
      status: 'DISPATCHED_200_OK',
      timestamp: new Date().toLocaleTimeString()
    }
  ];

  /**
   * Dispatches signed n8n webhooks for WHEN-THEN automation rules.
   */
  public static triggerEvent(
    eventType: 'CRITICAL_P1_ALERT' | 'SLA_UNACKNOWLEDGED_ESCALATION' | 'TICKET_RESOLVED_NOTIFY_USER',
    incident: Incident
  ): WebhookLogEntry {
    const ruleNameMap = {
      CRITICAL_P1_ALERT: 'Notify IT Manager (P1 Critical Alert)',
      SLA_UNACKNOWLEDGED_ESCALATION: 'Escalate Unacknowledged Ticket (>15m)',
      TICKET_RESOLVED_NOTIFY_USER: 'Send Resolution Email to User'
    };

    const endpointMap = {
      CRITICAL_P1_ALERT: 'http://localhost:5678/webhook/p1-critical-alert',
      SLA_UNACKNOWLEDGED_ESCALATION: 'http://localhost:5678/webhook/sla-escalation',
      TICKET_RESOLVED_NOTIFY_USER: 'http://localhost:5678/webhook/user-resolution-notify'
    };

    const payloadObj = {
      event: eventType,
      ticket: incident.ticketNumber,
      title: incident.title,
      severity: incident.severity,
      reporter: incident.reporter,
      timestamp: new Date().toISOString()
    };

    const entry: WebhookLogEntry = {
      id: `wh-${Date.now()}`,
      eventType,
      ruleName: ruleNameMap[eventType],
      targetEndpoint: endpointMap[eventType],
      payloadJson: JSON.stringify(payloadObj),
      hmacSignature: `sha256=${Math.random().toString(16).substring(2, 18)}...`,
      status: 'DISPATCHED_200_OK',
      timestamp: new Date().toLocaleTimeString()
    };

    this.liveStreamLogs.unshift(entry);
    return entry;
  }

  public static getLogs(): WebhookLogEntry[] {
    return this.liveStreamLogs;
  }
}
